### ISSUE TITLE
Be able to create dynamic fmt args array

### ISSUE THREAD

**Issue by Alexays** [Alexays/Waybar#100]
In order to avoid any unnecessary logic.
e.g. the CPU module without label {usage}, still calculates the usage when the argument is not provided

**Comment by Alexays** [comment]
Done for the CPU module in 2d9bcb1a2df916e4a6846659ce91c5cd8038097e

**Comment by Organic-Code** [comment]
I'll try to work on that

**Comment by Organic-Code** [comment]
Alright, so my idea was to implement some kind of lazy argument support, so that you could be able to do something like
```cpp
fmt::format(str, fmt::arg("data1", lazy(computeData1)), …)
```
where computeData1 would be called only when `"data1"` is present is the `str` format specifier
That part is very simple to do, but I find it quite cumbersome to use, especially when parameters are to be taken into account, leading to code looking like this:
```cpp
fmt::format(str, fmt::arg("data1", lazy([&] () mutable { return this->computeData1(param1, param2); }), …)
```
which I personally find pretty awful, especially where there would be several named arguments.


So I figured out something like this would be much nicer to use (taking the same function as above):
```cpp
fmt::format(str, lazy::arg("data1", this, computeData1, param1, param2), …)
```
This is however mostly not feasible with fmt as it is due to life time management issues.
If you agree with it, I'd like to modify fmt so as to be able to do this.
If however you disagree with that, then I'll try to go and improve the first method usage

**Comment by Alexays** [comment]
Yeah thats why 2d9bcb1a2df916e4a6846659ce91c5cd8038097e sigsev cause of the lifetime 
 of arguments.
I don't really want to have to modify fmt to implement this knowing that they will probably one day propose a way to make dynamic arguments.
Maybe we can override fmt::arg to do this in a cleaner way?

**Comment by Organic-Code** [comment]
I've tried various things, but unfortunately that's not possible AFAIK
Fmt::arg relies on the type `named_arg` to detect the use of named args, and  `named_arg` is the type storing a reference, which in turn causes the lifetime issues

**Comment by Organic-Code** [comment]
I just realized we could achieve a similar syntax without changing Fmt:
```cpp
fmt::format(str, fmt::arg("data1", lazy(this, computeData1, param1, param2)) , …)
```
Which is slightly worse, but is still OK IMHO

Well, using macros would actually allow the syntax I proposed with  `lazy::arg` by expanding to this code, but I don't think that would be a really good idea

**Comment by Alexays** [comment]
@Organic-Code I looked at it and without modifying fmt I think it's the best way to do it. Macro is not very useful in this case

**Comment by Organic-Code** [comment]
I'll work forward with that then

**Comment by 0xdec** [comment]
In working on #396 I realized that formatting is incredibly inconsistent between modules. I propose that each module should automatically recalculate its format and tooltip_format every time the state changes, and update its text and tooltip_text every time an argument changes. Also, more info should be made available as args (e.g. pulseaudio should have `desc` available as an arg). This would allow for a generic system that works across every module as long as each module publishes its relevant states and args.

**Comment by Alexays** [comment]
@0xdec I'm on it, I'll continue when I'll have some time :)

**Comment by 0xdec** [comment]
This would also enable some cool stuff that I describe in #402. I might have time to start looking at this, but idk.
